### PR TITLE
PRMT-4074

### PR DESCRIPTION
**Pull request by martin-nhs**

### CODE REVIEW

File: src/__tests__/app.integration.test.js [hunk 1]
```diff
@@ -273,30 +273,29 @@ describe('Ensure health record outbound XML is unchanged', () => {
...
     const fragmentMessageIds = Object.keys(originalFragments);
-    for (let messageId of fragmentMessageIds) {
-      // add records of the old message ids to database table
-      // new message ids are mostly same as the old ones, with last char replaced as '0', in order to guarantee the .sort() at expect statement give the same order.
-      await createMessageIdReplacement(messageId, messageId.slice(0, 35) + '0')
+    const messageIdReplacements = fragmentMessageIds.map(oldMessageId => {
+      return {
+        oldMessageId,
+        newMessageId: oldMessageId.slice(0, 35) + '0'
+      }
```
Comment (MohammadIqbalAD-NHS) [inline src/__tests__/app.integration.test.js:282]: Is there a blank line here or is GitHub just displaying it wrongly?

File: src/__tests__/db.integration.test.js [hunk 2]
```diff
@@ -1,115 +1,41 @@
...
-import { createRegistrationRequest } from '../services/database/create-registration-request';
-import { logger } from "../config/logging";
-import { createMessageIdReplacement } from "../services/database/create-message-id-replacement";
-import nock from "nock";
-import { config } from "../config";
-
-jest.mock('../services/ehr-repo/get-fragment');
-jest.mock('../services/transfer/transfer-out-util');
+import { createRegistrationRequest } from "../services/database/create-registration-request";
+import {logInfo} from "../middleware/logging";
```
Comment (MohammadIqbalAD-NHS) [inline src/__tests__/db.integration.test.js:5]:
```suggestion
import { logInfo } from "../middleware/logging";
```
Space between the curly brackets and variable name for consistency and readabilitiy

File: src/__tests__/db.integration.test.js [hunk 3]
```diff
@@ -1,115 +1,41 @@
...
-    for (let oldFragmentMessageId of fragmentMessageIds) {
-      const newFragmentMessageId = oldFragmentMessageId.slice(0, 35) + '0';
-      await createMessageIdReplacement(oldFragmentMessageId, newFragmentMessageId);
-      getFragment.mockResolvedValueOnce((JSON.parse(fragmentsWithMessageIds[oldFragmentMessageId])));
-      updateFragmentMessageId.mockResolvedValueOnce({
-        newMessageId: newFragmentMessageId ,
-        message: JSON.parse(fragmentsWithMessageIds[oldFragmentMessageId])
-      });
+    for (let i = 0; i < numberOfTransactions; i++) {
+      databaseOperationPromises.push(createRegistrationRequest(uuid().toUpperCase(), uuid().toUpperCase(), nhsNumber, odsCode));
```
Comment (MohammadIqbalAD-NHS) [inline src/__tests__/db.integration.test.js:29]:
```suggestion
      databaseOperationPromises.push(createRegistrationRequest(conversationId, messageId, nhsNumber, odsCode));
```
I suggest we create variables for `conversationId `and `messageId `and refer to them within this line to remove the ambiguity of what we are passing into the `createRegistrationRequest `function.

File: src/__tests__/message-id.integration.test.js [hunk 4]
```diff
@@ -13,10 +13,12 @@ import { readFileSync } from 'fs';
 import expect from "expect";
 import nock from 'nock';
 import { sortBy } from 'lodash';
+import { replaceMessageIdsInObject } from "../services/transfer/transfer-out-util";
 
 describe('Replacement of message IDs', () => {
-  // ============ COMMON PROPERTIES ============
-  const gp2gpUrl = 'http://fake-gp2gpmessager-url';
+  const uuidRegexPattern = /^[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}$/
```
Comment (MohammadIqbalAD-NHS) [inline src/__tests__/message-id.integration.test.js:19]: `uuidRegexPattern `is being declared on both Line 19 & Line 90. Remove one or the other.

File: src/__tests__/message-id.integration.test.js [hunk 5]
```diff
@@ -122,61 +122,57 @@ describe('Replacement of message IDs', () => {
...
   describe('Message fragments use new message IDs', () => {
-    function makePresignedUrlForFragment(messageId) {
-      return `http://fake-fragment-presign-url/${messageId}`;
+    function readMessageFragmentFile(fragmentFilename) {
+      return readFileSync(`src/__tests__/data/ehr_with_fragments/${fragmentFilename}`, 'utf8');
     }
 
-    function setUpMockForEhrRepoFragmentMessage(messageId) {
-      const ehrFragmentPresignedUrl = makePresignedUrlForFragment(messageId);
+    const getPresignedUrlForFragment = (messageId) => `http://fake-fragment-presign-url/${messageId}`;
```
Comment (MohammadIqbalAD-NHS) [inline src/__tests__/message-id.integration.test.js:141]:
```suggestion
    const getPresignedUrlForFragment = (messageId) => `http://fake-fragment-presigned-url/${messageId}`;
```

File: src/__tests__/message-id.integration.test.js [hunk 6]
```diff
@@ -190,69 +165,69 @@ describe('Replacement of message IDs', () => {
...
+        expect(fragmentOldMessageIds).not.toContain(postRequestBody.messageId);
+      });
 
-      for (let postBody of gp2gpMessengerPostBodies) {
-        expect(postBody.conversationId).toEqual(conversationId);
-        expect(postBody.odsCode).toEqual(odsCode);
+      // Get the new Message IDs from the database, compare with the
+      // Message IDs within the POST Request bodies.
+      const messageIdReplacements = await getAllMessageIdReplacements(fragmentOldMessageIds);
+      const newMessageFragmentIds = messageIdReplacements
```
Comment (MohammadIqbalAD-NHS) [inline src/__tests__/message-id.integration.test.js:188]:
```suggestion
      const newFragmentMessageIds = messageIdReplacements
```

File: src/__tests__/app.integration.test.js (hunk 1 above)
Comment (martin-nhs) [reply src/__tests__/app.integration.test.js:282]: GitHub glitch

File: src/__tests__/message-id.integration.test.js (hunk 4 above)
Comment (martin-nhs) [reply src/__tests__/message-id.integration.test.js:17]: Fixed

File: src/services/transfer/transfer-out-util.js [hunk 7]
```diff
@@ -51,70 +52,32 @@ export const updateFragmentStatus = async (conversationId, messageId, status) =>
...
-export const updateMessageIdForEhrCore = async ehrCore => {
-  try {
-    const messageId = await parseMessageId(ehrCore);
-    const newMessageId = uuidv4().toUpperCase();
-    const ehrCoreWithUpdatedMessageId = replaceMessageIdInObject(ehrCore, messageId, newMessageId);
+  return JSON.parse(ehrMessageJsonString);
+}
 
+export const createNewMessageIds = async oldMessageIds => {
+  const messageIdReplacements = oldMessageIds.map((oldMessageId) => {
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-util.js:67]:
```suggestion
  const messageIdReplacements = oldMessageIds.map(oldMessageId => {

```

File: src/services/transfer/transfer-out-util.js [hunk 8]
```diff
@@ -51,70 +52,32 @@ export const updateFragmentStatus = async (conversationId, messageId, status) =>
...
-};
+  });
 
-export const createNewMessageIdsForAllFragments = async fragmentMessageIds => {
-  for (let oldMessageId of fragmentMessageIds) {
-    const newMessageId = uuidv4().toUpperCase();
-    await createMessageIdReplacement(oldMessageId, newMessageId);
-  }
+  await createMessageIdReplacements(messageIdReplacements);
+  logInfo(`Created new message ID(s) for EHR core & ${messageIdReplacements.length - 1} fragments`);
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-util.js:75]: What if there was 1 fragment, would this be `Created new message ID(s) for EHR core & 0 fragments`?

File: src/services/database/__tests__/create-message-id-replacement.integration.test.js [hunk 9]
```diff
@@ -25,7 +25,8 @@ describe('createMessageIdReplacement', () => {
     const newMessageId = uuidv4();
 
     // when
-    await createMessageIdReplacement(oldMessageId, newMessageId);
+    await createMessageIdReplacements([{oldMessageId, newMessageId}]);
+    // await createMessageIdReplacements([1,2,3]);
```
Comment (MohammadIqbalAD-NHS) [inline src/services/database/__tests__/create-message-id-replacement.integration.test.js:29]: Commented out code.

File: src/services/transfer/transfer-out-util.js (hunk 8 above)
Comment (martin-nhs) [inline src/services/transfer/transfer-out-util.js:75]:
```suggestion
  logInfo(`Created new Message ID's for EHR Core and ${messageIdReplacements.length - 1} fragment(s)`);
```

File: src/services/transfer/transfer-out-fragments.js [hunk 10]
```diff
@@ -1,26 +1,101 @@
...
+
+const handleFragmentTransferError = async (error, conversationId, messageId) => {
+  switch (true) {
+    case error instanceof PresignedUrlNotFoundError:
+      await updateFragmentStatus(conversationId, messageId, Status.MISSING_FROM_REPO);
+      break;
+    case error instanceof DownloadError:
+      await updateFragmentStatus(conversationId, messageId, Status.DOWNLOAD_FAILED);
+      break;
+    // this will catch any miscellaneous errors
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-fragments.js:96]: Redundant comment - can be removed, nature of `default` is to catch any unmatching criteria.

File: src/services/transfer/transfer-out-fragments.js [hunk 11]
```diff
@@ -1,26 +1,101 @@
...
+  });
+
+  logInfo(`Out of ${newMessageIds.length} message Ids, ` +
+  `${newMessageIds.length - messageIdsOfFragmentsEligibleForSending.length} have already been sent. ` +
+  `${messageIdsOfFragmentsEligibleForSending.length} are eligible to be sent`);
+
+  return messageIdsOfFragmentsEligibleForSending;
+}
+
+const isFragmentIsEligibleToBeSent = (messageFragmentRecord) => {
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-fragments.js:82]: Typo in function: 

```suggestion
const isFragmentEligibleToBeSent = (messageFragmentRecord) => {
```

File: src/services/transfer/transfer-out-fragments.js [hunk 12]
```diff
@@ -1,26 +1,101 @@
...
+          await new Promise(executor => setTimeout(executor, config.fragmentTransferRateLimitTimeoutMilliseconds));
+      } catch (error) {
+        await handleFragmentTransferError(error, conversationId, messageIdWithReplacement.newMessageId);
+        throw error;
+      }
+    }
+    logInfo(`All fragments have been successfully sent to GP2GP Messenger.`);
+}
+
+const getMessageIdsOfFragmentsEligibleForSending = async (messageIdsWithReplacements) => {
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-fragments.js:64]:
```suggestion
const getMessageIdsForAllFragmentsEligibleForSending = async messageIdsWithReplacements => {

```

File: src/services/handler/__test__/continue-message-handler.test.js [hunk 13]
```diff
@@ -1,14 +1,22 @@
...
 import { parseConversationId } from "../../parser/parsing-utilities";
 import continueMessageHandler from "../continue-message-handler";
-import { NhsNumberNotFoundError } from "../../../errors/errors";
 import { Status } from "../../../models/registration-request";
 import { readFileSync } from "fs";
 import expect from "expect";
 import path from "path";
+import {logError, logInfo, logWarning} from "../../../middleware/logging";
+import {hasServiceStartedInTheLast5Minutes} from "../../../config";
+import {NhsNumberNotFoundError} from "../../../errors/errors";
```
Comment (MohammadIqbalAD-NHS) [inline src/services/handler/__test__/continue-message-handler.test.js:19]: Space between curly brackets and variable names for consistency/readability

File: src/services/handler/continue-message-handler.js [hunk 14]
```diff
@@ -14,6 +22,39 @@ export default async function continueMessageHandler(message) {
...
+      logWarning(`Ignoring duplicate continue request. Conversation ID ${conversationId} already completed successfully`);
+      break;
+
+    case Status.INCORRECT_ODS_CODE:
+    case Status.MISSING_FROM_REPO:
+    case Status.EHR_DOWNLOAD_FAILED:
+    case Status.CORE_SENDING_FAILED:
+    case Status.EHR_INTEGRATION_FAILED:
+      logWarning(`Ignoring duplicate continue request. Conversation ID ${conversationId} already failed and is unable to retry`);
+      break;
```
Comment (MohammadIqbalAD-NHS) [inline src/services/handler/continue-message-handler.js:39]: A bunch of cases with no actions. Intentional?

File: src/services/transfer/__tests__/transfer-out-fragments.test.js [hunk 15]
```diff
@@ -1,80 +1,266 @@
...
+  const outboundConversationId = 'cbc75815-8ee8-4554-980a-f7f8fd2cd341';
+
+  const originalMessageId1 = '2526ea9f-f800-461d-8046-8bc512920d23';
+  const originalMessageId2 = '6e7725e3-0cab-4fd8-a518-eeda7b5925a6';
+  const updatedMessageId1 = '7B2D6A16-2C50-4BF8-A923-E14003816ECC';
+  const updatedMessageId2 = 'DDA75C2B-18B6-424D-B94B-83205AAF2FEA';
+
+  const messageIdsWithReplacements = [
+    { oldMessageId: originalMessageId1, newMessageId: updatedMessageId1 },
+    { oldMessageId: originalMessageId2, newMessageId: updatedMessageId2 },
```
Comment (MohammadIqbalAD-NHS) [inline src/services/transfer/__tests__/transfer-out-fragments.test.js:45]:
```suggestion
    { oldMessageId: originalMessageId2, newMessageId: updatedMessageId2 }

```

File: src/services/transfer/__tests__/transfer-out-fragments.test.js [hunk 16]
```diff
@@ -1,80 +1,266 @@
...
+  }
+
   // =================== END ===================
 
   afterEach(() => {
     jest.resetAllMocks();
   });
 
-  it('should send fragment on success', async () => {
+  it('should transfer out fragments for a new continue request', async () => {
```
Comment (MohammadIqbalAD-NHS) [inline src/services/transfer/__tests__/transfer-out-fragments.test.js:83]:
```suggestion
  it('should transfer out fragments when new continue request received', async () => {

```

File: src/services/handler/continue-message-handler.js (hunk 14 above)
Comment (martin-nhs) [reply src/services/handler/continue-message-handler.js:39]: https://www.freecodecamp.org/news/fall-through-in-javascript-switch-statements/

File: src/services/transfer/__tests__/transfer-out-fragments.test.js [hunk 17]
```diff
@@ -1,80 +1,266 @@
...
   const inboundConversationId = 'a52e8782-2268-4d49-ab69-5fcfa10eb43a';
-  const originalMessageId = '2526ea9f-f800-461d-8046-8bc512920d23';
-  const updatedMessageId = '7B2D6A16-2C50-4BF8-A923-E14003816ECC';
-  const fragment = {ebXML: '', payload: '', attachments: [], externalAttachments: []};
+  const outboundConversationId = 'cbc75815-8ee8-4554-980a-f7f8fd2cd341';
+
+  const originalMessageId1 = '2526ea9f-f800-461d-8046-8bc512920d23';
+  const originalMessageId2 = '6e7725e3-0cab-4fd8-a518-eeda7b5925a6';
+  const updatedMessageId1 = '7B2D6A16-2C50-4BF8-A923-E14003816ECC';
+  const updatedMessageId2 = 'DDA75C2B-18B6-424D-B94B-83205AAF2FEA';
```
[the remark this answers is not in the document]
Comment (martin-nhs) [reply src/services/transfer/__tests__/transfer-out-fragments.test.js:41]: had discussions around this with @AndyFlintAnswerDigital and @MohammadIqbalAD-NHS - we've agreed that it's neutral feelings and we will proceed with this for the time being.

File: src/services/transfer/transfer-out-ehr-core.js [hunk 18]
```diff
@@ -68,39 +57,41 @@ export async function transferOutEhrCore({
...
-    logInfo(`EHR out transfer with conversation ID ${conversationId} is already in progress`);
+    logWarning(`EHR out transfer with conversation ID ${conversationId} is already in progress`);
     return true;
   }
+
   return false;
 };
+
+const handleCoreTransferError = async (error, conversationId) => {
+  switch (true) {
```
Comment (MohammadIqbalAD-NHS) [inline src/services/transfer/transfer-out-ehr-core.js:86]: First time I have seen `switch (true)` used before.
(O.o)

Comment (martin-nhs) [reply src/services/transfer/transfer-out-ehr-core.js:86]: As `error instanceof XYZError` evaluates to either true or false, we want to switch on true for the case to be run!

**Review by MohammadIqbalAD-NHS — APPROVED** [review]

File: src/services/database/__tests__/message-id-replacement-repository.integration.test.js [hunk 19]
```diff
@@ -1,73 +1,88 @@
...
+        oldMessageIds: getAllOldMessageIds(messageIdReplacementRecords),
+        newMessageIdsFromResult: getAllNewMessageIds(result),
+        newMessageIds: getAllNewMessageIds(messageIdReplacementRecords).map(messageId => messageId.toUpperCase())
+      }
 
       // then
-      expect(newMessageIdFromRecord).toBe(newMessageId);
-      expect(newMessageIdFromRecord).toBe(newMessageIdFromRecord.toUpperCase());
+      expect(messageIds["oldMessageIdsFromResult"]).toEqual(messageIds["oldMessageIds"]);
+      expect(messageIds["newMessageIdsFromResult"]).toEqual(messageIds["newMessageIds"]);
```
Comment (AndyFlintNHS) [inline src/services/database/__tests__/message-id-replacement-repository.integration.test.js:42]: Not necessary, but to be consistent with our other tests, let's change the syntax of these to `messageIds.newMessageIdsFromResult`

File: src/services/database/message-id-replacement-repository.js [hunk 20]
```diff
@@ -4,12 +4,24 @@ import { FragmentMessageIdReplacementRecordNotFoundError } from '../../errors/er
...
+      }
+    });
   });
-};
+}
+
+const verifyMessageIdReplacementWasFoundForEachMessageId = (oldMessageIds, messageIdReplacements) => {
+  if (messageIdReplacements.length !== oldMessageIds.length)
+    throw new FragmentMessageIdReplacementRecordNotFoundError(oldMessageIds.length, messageIdReplacements.length);
+}
```
Comment (AndyFlintNHS) [inline src/services/database/message-id-replacement-repository.js:27]: New line at the end of the file

File: src/services/database/registration-request-repository.js [hunk 21]
```diff
@@ -7,16 +7,17 @@ import { Op } from "sequelize";
 
 const RegistrationRequest = ModelFactory.getByName(modelName);
 
-export const getRegistrationRequestStatusByConversationId = conversationId => {
-  return RegistrationRequest.findByPk(conversationId);
+export const getRegistrationRequestByConversationId = async conversationId => {
+  return await RegistrationRequest.findByPk(conversationId);
 };
 
 export const getNhsNumberByConversationId = conversationId => {
```
Comment (AndyFlintNHS) [inline src/services/database/registration-request-repository.js:14]: Again, let's discuss the async

**Review by AndyFlintNHS — CHANGES_REQUESTED** [review]

File: src/services/database/registration-request-repository.js (hunk 21 above)
Comment (martin-nhs) [reply src/services/database/registration-request-repository.js:14]: To resolve on a separate ticket @AndyFlintAnswerDigital

File: src/services/database/create-message-id-replacements.js [hunk 22]
```diff
@@ -1,23 +1,19 @@
...
-      })
+import { runWithinTransaction } from './helper';
+import { logError, logInfo } from "../../middleware/logging";
+import ModelFactory from '../../models';
+import { modelName } from '../../models/message-id-replacement';
+import { errorMessages } from "../../errors/errors";
+
+const MessageIdReplacement = ModelFactory.getByName(modelName);
+
+export const createMessageIdReplacements = (messageIdReplacements) =>
```
[the remark this answers is not in the document]
Comment (martin-nhs) [reply src/services/database/create-message-id-replacements.js]: As the usage of this function uses `await` I will add `async` to this function, though should be reviewed and addressed as part of the action noted on https://github.com/nhsconnect/prm-repo-ehr-out-service/pull/65#discussion_r1368450392.

**Pull request merged**